src/backend/dx12/src/command.rs

```diff
@@ -691,13 +691,28 @@ impl CommandBuffer {
         offset: u32,
         constants: &[u32],
     ) {
-        let num = constants.len() as u32;
+        assert!(offset % 4 == 0);
+
+        let pc_start = offset / 4;
+        let pc_end = pc_start + constants.len() as u32;
+
         for root_constant in &layout.root_constants {
-            assert!(root_constant.range.start <= root_constant.range.end);
-            if root_constant.range.start >= offset && root_constant.range.start < offset + num {
-                let start = (root_constant.range.start - offset) as _;
-                let end = num.min(root_constant.range.end - offset) as _;
-                user_data.set_constants(offset as _, &constants[start..end]);
+            debug_assert!(root_constant.range.start <= root_constant.range.end);
+
+            let intersection =
+                root_constant.range.start.max(pc_start)..root_constant.range.end.min(pc_end);
+
+            if intersection.len() > 0 {
+                debug_assert!(intersection.start <= intersection.end);
+                debug_assert!(intersection.start >= pc_start);
+                debug_assert!(intersection.start >= root_constant.range.start);
+
+                let pc_start = (intersection.start - pc_start) as usize;
+                let pc_end = pc_start + intersection.len();
+
+                let rc_start = (intersection.start - root_constant.range.start) as usize;
+
+                user_data.set_constants(rc_start, &constants[pc_start..pc_end]);
```

Comment:
Should be `intersection.start` instead of `rc_start` I think

Reply:
I think you're right. I'll take a look at this tomorrow.

Reply:
I feel like I'm missing something. If we don't need to know the root constant range here at all, there's no reason this entire function can't be just the following? Could it really be that simple?

```rust
fn push_constants(
    user_data: &mut UserData,
    layout: &r::PipelineLayout,
    offset: u32,
    constants: &[u32],
) {
    assert!(offset % 4 == 0);
    user_data.set_constants(offset as usize / 4, &constants);
}
```

Reply:
Actually, I think that's valid as we delayed the actual pushing of the root constants.

Reply:
I went ahead and changed it.